zkevm-circuits/src/evm_circuit/step.rs

```diff
@@ -322,10 +323,96 @@ impl ExecutionState {
     }
 }
 
+/// Dynamic selector that generates expressions of degree 2 to select from N
+/// possible targets using N/2 + 1 cells.
+#[derive(Clone, Debug)]
+pub(crate) struct DynamicSelectorHalf<F> {
+    /// Whether the target is even.  `target % 2 == 0`.
+    pub(crate) target_even: Cell<F>,
+    /// Whether the target belongs to each consecutive pair of targets.
+    /// `in [0, 1], in [2, 3], in [4, 5], ...`
+    pub(crate) target_pairs: Vec<Cell<F>>,
+}
+
+impl<F: FieldExt> DynamicSelectorHalf<F> {
+    pub(crate) fn new(cell_manager: &mut CellManager<F>, count: usize) -> Self {
+        let target_pairs = cell_manager.query_cells(CellType::Storage, (count + 1) / 2);
+        let target_even = cell_manager.query_cell(CellType::Storage);
+        Self {
+            target_pairs,
+            target_even,
+        }
+    }
+
+    /// Return the list of constraints that configure this "gadget".
+    pub(crate) fn configure(&self) -> Vec<(&'static str, Expression<F>)> {
+        // Only one of target_pairs should be enabled
+        let sum_to_one = (
+            "Only one of target_pairs should be enabled",
+            self.target_pairs
+                .iter()
+                .fold(1u64.expr(), |acc, cell| acc - cell.expr()),
+        );
+        // Cells representation for target_pairs and target_even should be bool.
+        let bool_checks = iter::once(&self.target_even)
+            .chain(&self.target_pairs)
+            .map(|cell| {
+                (
+                    "Representation for target_pairs and target_even should be bool",
+                    cell.expr() * (1u64.expr() - cell.expr()),
+                )
+            });
+        iter::once(sum_to_one).chain(bool_checks).collect()
+    }
```

Comment:
It seems we need to disallow the invalid case just like `BinaryNumberChip`, to avoid invalid selector be enabled all the way down to escape constraints?

If so in this case we only need to add an extra constraint `target_even * target_pairs.last() == 0` when `count` is odd.

Reply:
Well spotted!  Thanks!  I've fixed this in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/718/commits/e2bcc55aea01340382f818e3c6d7c88a73aac98e

I've also replaced the usage of `even` flag by `odd` flag so that it matches with the less significant bit of the value.  I think this way it's easier to reason about.

Reply:
Makes sense!